relayer/processor/path_processor.go

```diff
@@ -168,33 +179,60 @@ func (pp *PathProcessor) HandleNewData(chainID string, cacheData ChainProcessorC
 	}
 }
 
+func (pp *PathProcessor) processAvailableSignals(ctx context.Context, cancel func(), messageLifecycle MessageLifecycle) error {
+	select {
+	case <-ctx.Done():
+		pp.log.Debug("Context done, quitting PathProcessor",
+			zap.String("chain_id_1", pp.pathEnd1.info.ChainID),
+			zap.String("chain_id_2", pp.pathEnd2.info.ChainID),
+			zap.String("client_id_1", pp.pathEnd1.info.ClientID),
+			zap.String("client_id_2", pp.pathEnd2.info.ClientID),
+			zap.Error(ctx.Err()),
+		)
+		return ctx.Err()
+	case d := <-pp.pathEnd1.incomingCacheData:
+		// we have new data from ChainProcessor for pathEnd1
+		pp.pathEnd1.MergeCacheData(ctx, cancel, d, messageLifecycle)
+
+	case d := <-pp.pathEnd2.incomingCacheData:
+		// we have new data from ChainProcessor for pathEnd2
+		pp.pathEnd2.MergeCacheData(ctx, cancel, d, messageLifecycle)
+
+	case <-pp.retryProcess:
+		// No new data to merge in, just retry handling.
+	}
+	return nil
+}
+
 // Run executes the main path process.
-func (pp *PathProcessor) Run(ctx context.Context) {
+func (pp *PathProcessor) Run(ctx context.Context, cancel func(), messageLifecycle MessageLifecycle) {
+	var retryTimer *time.Timer
 	for {
-		select {
-		case <-ctx.Done():
+		// block until we have any signals to process
+		if err := pp.processAvailableSignals(ctx, cancel, messageLifecycle); err != nil {
```

Comment:
`processAvailableSignals` logs an error and returns that error, or it returns nil. We discard the error here. So this may read better as checking a boolean, or at least leave a comment inline saying why we discard the error instead of logging it.

Reply:
changed to a bool